Add system settings manager

ArturoManzoli opened the pull request:
* Users can edit, save and upload settings;
* Reset settings button will be implemented on a second approach. For now a reset for the browser storage is implemented only.

https://github.com/user-attachments/assets/ab2a4507-b1c5-4ee1-9267-2b4a132c0072

Closes #1486

rafaellehmkuhl commented on src/assets/defaults.ts:
```diff
@@ -901,3 +907,90 @@ export const defaultSensorDataloggerProfile: OverlayGrid = {
   distance: DistanceDisplayUnit.Meters,
 }
+
+export const defaultCockpitSettings = {
```
I liked the idea of having all the defaults centralized in one place!

One thing thought is that we need to use those in the places where they are required (e.g.: calls to `useStorage` and `useBlueOsStorage`), otherwise if someone changes the value there, the value here will be deprecated and the code will not work.

rafaellehmkuhl commented on src/assets/defaults.ts:
```diff
@@ -901,3 +907,90 @@ export const defaultSensorDataloggerProfile: OverlayGrid = {
+    { level: 'warning', enabled: true },
+    { level: 'critical', enabled: true },
+  ],
+  'cockpit-mission-start-time': '2025-02-10T21:34:50.113Z',
```
It's crucial that we don't change the default values from what they are today. This one should be `new Date()`, and not that random date.

The same happens for several others. I believe you copied your own settings instead of the real defaults.

rafaellehmkuhl commented on src/components/configuration/ManageCockpitSettings.vue:
```diff
@@ -0,0 +1,425 @@
+
+// Trying to sync automatically
+watch(
+  () => userSettings.value,
```
A tip here: it looks like you're creating an extra localStorage entry to store a duplication of all variables just to be alerted by the `useStorage` composable about changes in the value. You can instead just listen to the "storage" event. This way you don't duplicate everything (which can cause a lot of confusion for someone reading the stored values, as there would be two possible two sources of truth) and you can monitor for changes directly in the values of the keys.

```javascript
addEventListener("storage", (event) => {});
```

Actually, from this code, I assume you're not receiving changes made to the original settings, right?

ArturoManzoli replied:
Correct, I'm not receiving the changes. 

Actually what I was trying to do there was to check for changes on the local copy of the settings. and if there was any change, save it to the storage.

I'll implement the event listener and check the results

rafaellehmkuhl commented on src/views/ConfigurationGeneralView.vue:
```diff
@@ -519,7 +533,6 @@ const tryToPrettifyRtcConfig = (): void => {
 
 const openTutorial = (): void => {
   interfaceStore.isMainMenuVisible = false
-  interfaceStore.configComponent = -1
```
To better understand, why was this removed?

ArturoManzoli replied:
This should be on its own commit, I believe.

On a recent PR you added a 'tools' submenu that modified a little bit the way submenus and its contents work. This was leftover from that, I suppose.

![image](https://github.com/user-attachments/assets/cf10617f-0e4c-439c-80a2-a6c733735e55)

rafaellehmkuhl replied:
Got it! Was a leftover indeed!

Yes, makes sense to be on a separate PR.

Sad that our linter is partially broken and not catching all errors. I should revisit the PR that fixes it at some point.

rafaellehmkuhl replied:
Sorry, I meant commit and not PR 😅

rafaellehmkuhl commented on src/components/configuration/CockpitSettingsManager.vue:
```diff
@@ -0,0 +1,400 @@
+        text: 'Reset settings',
+        action: () => {
+          localStorage.clear()
+          location.reload()
```
This will reload the page immediately, so the snackbar won't show or will do for a quick moment.

There's a utils method that deals with warning about the reload and setting a timeout so the user can read any message left. It can help here.

ArturoManzoli commented on src/views/ConfigurationGeneralView.vue:
```diff
@@ -16,35 +16,46 @@
                   Show tutorial
                 </v-btn>
                 <v-btn
-                  v-if="isElectron()"
```
That was a mistake I made. Nice that you saw. I was checking the menu with all buttons and forgot to put this back. 

Returning now. Thanks

rafaellehmkuhl reviewed (commented):
Code looks good now. 
Will just test it before approve.

rafaellehmkuhl reviewed (commented):
Could we have an edit button in the right to enable editing the value? Or maybe enabling it with a single click? Having to double-click requires the user to guess what to do. It's also not working on touch/mobile.

ArturoManzoli commented:
> Could we have an edit button in the right to enable editing the value? Or maybe enabling it with a single click? Having to double-click requires the user to guess what to do. It's also not working on touch/mobile.

Done!

rafaellehmkuhl reviewed (changes requested):
I got somewhat confused with the behavior of the edit button. When I click it, it says "Done" and puts a green checkmark. It seems like something was completed. Strange and unexpected for an edit button. Is the checkmark like a "save" button?

https://github.com/user-attachments/assets/c1473800-1a54-43c3-8a3f-e91833e5052d

It's also unexpected to have different behaviors in those terms between the JSON variables and the primitive ones. Could we instead have the same for both?

A suggestion would be:

- Instead of the "Edit json" button, put a "{ ... }" there, as this was supposed to be the value, and we are not showing just because it doesn't fit there
- Use a pencil icon to start editing in all cases (json and not-json)
- Have the "Close" and "Save" buttons as the way to undo and save for all cases (json or primitive)
- Add a visual hint for the string and number values, as currently they don't seem to have entered an edit mode (different from the boolean and json ones, where this is clear)

ArturoManzoli commented:
On last patch:
* Added  'save' and 'cancel' buttons for primitives editing;
* Made more clear when a primitive value is being edited;
* Positioned the tooltips over the top of their parents;
* Fixed null variables not opening the input area (check last-mission-name and username values)

https://github.com/user-attachments/assets/ef63a892-e411-48d3-ac26-bd1ab61a3c09

rafaellehmkuhl reviewed (changes requested):
For some reason it is showing this old "main-menu-style" variable (which was replaced by "cockpit-main-menu-style"), even with it not being in my `localStorage` anymore.

Also, if i change any other variable, this old "main-menu-style" variable is being recreated.

<img width="1496" alt="image" src="https://github.com/user-attachments/assets/9b50c72d-3f28-4352-8bdd-3e9ac1804b64" />

.

Could we also add a clear button in the search input?

ArturoManzoli commented:
> For some reason it is showing this old "main-menu-style" variable (which was replaced by "cockpit-main-menu-style"), even with it not being in my `localStorage` anymore.
> 
> Also, if i change any other variable, this old "main-menu-style" variable is being recreated.
>  
> Could we also add a clear button in the search input?

All variables had its prefix "cockpit-" trimmed out for display only purposes. Users don't need to read all those redundant prefixes, since the important part for them is what comes after it.
But anyway the variable being saved has the prefix on it.

I couldn't replicate the main-menu-style duplicating bug you mentioned and showed on the screen-capture.

Clear button added to the search bar!!

![image](https://github.com/user-attachments/assets/5cf57571-097d-4fd8-bdc3-5e7c4e2f6e4e)

rafaellehmkuhl commented:
> > For some reason it is showing this old "main-menu-style" variable (which was replaced by "cockpit-main-menu-style"), even with it not being in my `localStorage` anymore.
> > Also, if i change any other variable, this old "main-menu-style" variable is being recreated.
> > Could we also add a clear button in the search input?
> 
> All variables had its prefix "cockpit-" trimmed out for display only purposes. Users don't need to read all those redundant prefixes, since the important part for them is what comes after it. But anyway the variable being saved has the prefix on it.
> 
> I couldn't replicate the main-menu-style duplicating bug you mentioned and showed on the screen-capture.
> 
> Clear button added to the search bar!!

The bug is still there:

https://github.com/user-attachments/assets/09190bea-0bdd-45bb-b7a4-b08cc3511ca0

ArturoManzoli commented:
> The bug is still there:

Thanks for pointing that out! 
Now the bug it should be solved

On last patch:
* Removed redundant settings source;
* Added an event listener that check for settings changes made elsewhere on the system, and updates the settings manager content.

rafaellehmkuhl reviewed (approved):
Nice!

It's working fine now!
